### PR TITLE
Overhaul the text parsers, port from `nom` to `winnow`

### PR DESCRIPTION
**Pull request by zslayton**
This PR migrates the text parser from [`nom`](https://docs.rs/nom/latest/nom/) to an actively developed fork called [`winnow`](https://docs.rs/winnow/). You can read about the relationship between the two [here](https://docs.rs/winnow/latest/winnow/_topic/why/index.html#nom).

The migration offers a number of benefits:
* `winnow` includes a `debug` feature that lets you see the reader's path through the branches of the parser, making it MUCH easier to find the root of parser misbehavior.
* Where `nom` offers two flavors of each parser, `complete` and `streaming`, `winnow` allows an input source to report whether it is partial or complete. This means that there is a single flavor of each parser that will do the correct thing when it runs out of data. This allowed me to remove a LOT of special cases.
* The `Parser` trait takes `&mut self` and modifies it in-place rather than taking `self` and returning an updated copy along with the expected output value. (Discussion [here](https://docs.rs/winnow/latest/winnow/_topic/why/index.html#design-trade-offs).) Reducing the return value of every method from `(TextBuffer<'_>, T)` to just `T` increases the odds that the value will be returned in a register. Happily, because `TextBuffer` is `Copy`, we can still make as many intermediate state copies as we'd like when it's called for.
* The `alt((...))` combinator tries each of the provided parsers in turn and takes the first match. This is often quite slow--`winnow` provides [a `dispatch!` macro](https://docs.rs/winnow/latest/winnow/combinator/macro.dispatch.html) that allows you to prune the tree of options up-front by `match`ing on the head of the stream.
* Some simple types are now considered parsers themselves, which makes the parser methods that use them easier to read. For example, `tag("foo")`/`literal("foo")` can now be expressed as just `"foo"`. Similarly, tuples of parsers are now themselves parsers, so you no longer need to write `tuple(("/*", multiline_body_comment, "*/"))`. You can just write `("/*", multiline_body_comment, "*/")`.

I also made several improvements that did not require `winnow` per se:
* Made several encoding-version-specific methods and types generic over `E: TextEncoding`, eliminating a large amount of mostly duplicated code.
* Made a few parsers do an easily inlineable up-front check before calling the real (not as inlineable) complex implementation.
* Modified the 1.0 container parsers to cache the sub-expressions they encounter during lexing using the bump allocator, offering a big speedup. (This optimization had already been applied to the 1.1 container parsers.)

Hopefully this makes it much easier to both read and maintain.

--------

### Performance improvements

image

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

---------

### Testing improvements

Because nearly all of the tests read some amount of Ion text, this patch lead to a huge drop in the time needed to run the test harness.

#### Command
This command runs everything but the doc tests:
```
time cargo test --all-features --lib --tests
```

#### Before
image

#### After
image

### CODE REVIEW

File: src/lazy/text/buffer.rs [hunk 1]
```diff
@@ -3060,11 +2384,9 @@ mod tests {
             "_123",        // Leading underscore
             "0x0x5",       // Multiple 0x prefixes
             "0xx5",        // Multiple Xs after 0
-        ],
-        expect_incomplete: [
             "0x",          // Base 16 prefix w/no number
             "0b",          // Base 2 prefix w/no number
-        ]
+        ],
```
Comment (zslayton) [inline src/lazy/text/buffer.rs:2389]: 🪧 Because these unit tests are all reading from fixed slices, the parser will never return `Incomplete`. Those inputs have been moved to the `expect_mismatch` sections. We have [a separate test suite](https://github.com/amazon-ion/ion-rust/blob/89d3206c050666877494e0bb7364b041f504c0f5/tests/detect_incomplete_text.rs#L67-L69) just for incompleteness detection anyway.

File: src/lazy/text/raw/sequence.rs [hunk 2]
```diff
@@ -163,106 +139,109 @@ impl<'data> Iterator for RawTextListIterator_1_0<'data> {
...
+impl<'data, E: TextEncoding<'data>> LazyRawContainer<'data, E> for RawTextSExp<'data, E> {
+    fn as_value(&self) -> <E as Decoder>::Value<'data> {
         self.value
     }
 }
 
-impl<'data> LazyRawSequence<'data, TextEncoding_1_0> for LazyRawTextSExp_1_0<'data> {
-    type Iterator = RawTextSExpIterator_1_0<'data>;
+impl<'data, E: TextEncoding<'data>> LazyRawSequence<'data, E> for RawTextSExp<'data, E> {
+    type Iterator = RawTextSequenceCacheIterator<'data, E>;
```
Comment (zslayton) [inline src/lazy/text/raw/sequence.rs:244]: 🪧 All of the text containers now cache their child expressions and iterate over the cache as needed.

File: src/lazy/text/raw/sequence.rs [hunk 3]
```diff
@@ -87,67 +88,42 @@ impl Debug for LazyRawTextList_1_0<'_> {
...
         }
-        match self.input.match_list_value() {
-            Ok((remaining, Some(value))) => {
-                self.input = remaining;
-                let value = RawValueExpr::ValueLiteral(LazyRawTextValue_1_0::from(value));
-                Some(Ok(value))
-            }
-            Ok((_remaining, None)) => {
+        let result = whitespace_and_then(alt((
+            "]".value(None),
```
Comment (popematt) [inline src/lazy/text/raw/sequence.rs:115]: Should this one be `peek("]").value(None)` like the equivalent code for S-expressions?

File: src/lazy/binary/raw/reader.rs [hunk 4]
```diff
@@ -272,7 +283,9 @@ mod tests {
             [1, true, foo]
         "#,
         )?;
-        let mut reader = LazyRawBinaryReader_1_0::new(data);
+        let context = EncodingContext::empty();
+        let context = context.get_ref();
+        let mut reader = LazyRawBinaryReader_1_0::new(context, data);
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/binary/raw/reader.rs:288]: Sure, done!

File: src/lazy/binary/raw/reader.rs [hunk 5]
```diff
@@ -70,10 +78,8 @@ impl<'data> LazyRawBinaryReader_1_0<'data> {
         Ok(RawStreamItem::Value(lazy_value))
     }
 
-    pub fn next<'top>(&'top mut self) -> IonResult<LazyRawStreamItem<'top, BinaryEncoding_1_0>>
-    where
-        'data: 'top,
-    {
+    #[allow(clippy::should_implement_trait)]
+    pub fn next(&mut self) -> IonResult<LazyRawStreamItem<'data, BinaryEncoding_1_0>> {
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/binary/raw/reader.rs:82]: I wish! This would require [lending iterators](https://rust-lang.github.io/generic-associated-types-initiative/design_patterns/iterable.html), which are not provided by the standard library (yet?).

Long story short: the current `Iterator` trait returns values with the same lifetime as the thing over which you are iterating. We need an iterator that returns values that live only for the duration of that step of the iteration, so you can't get the next value until you drop the current one. In our case, we need a constrained lifetime because the encoding context can change between top-level values; we can't give out a lazy value and risk having the encoding context changed out from under its feet when we get the next one.

I've been mulling an API between `LazyValue` and `Element` where each value holds an `Rc` handle to both the encoding context and the byte buffer to which it refers, granting them a static lifetime in exchange for some bookkeeping overhead. It might even be possible to do as a shim over `LazyValue` on an opt-in basis, but I'd need to experiment.

Note that the readers _do_ have an `elements()` iterator that yields `Element`s, which is nice for use cases that don't mind the resource usage.

File: src/lazy/streaming_raw_reader.rs [hunk 6]
```diff
@@ -118,12 +128,18 @@ impl<Encoding: Decoder, Input: IonInput> StreamingRawReader<Encoding, Input> {
         self.read_next(context, /*is_peek=*/ true)
     }
 
+    fn input_is_streaming(&self) -> bool {
+        unsafe { &*self.input.get() }.is_streaming()
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/streaming_raw_reader.rs:132]: Good call--I've done you one better! By taking `&mut self` (even though this doesn't mutate anything), I can statically guarantee that this is safe to call. It's no longer `unsafe`.

[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/streaming_raw_reader.rs:132]: Make that two better! Using this trick (only on private methods), I was able to remove all of the `unsafe`s in the `StreamingRawReader` but one. 🙌

File: src/lazy/text/matched.rs [hunk 7]
```diff
@@ -1235,8 +1233,8 @@ mod tests {
             let expected: Int = expected.into();
             let encoding_context = EncodingContext::empty();
             let context = encoding_context.get_ref();
-            let buffer = TextBuffer::new(context, data.as_bytes());
-            let (_remaining, matched) = buffer.match_int().unwrap();
+            let buffer = TextBuffer::new(context, data.as_bytes(), true);
+            let matched = buffer.clone().match_int().unwrap();
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/text/matched.rs:1237]: It's a weird corner case for the unit tests. 

Winnow's parsers use `&mut self` and consume any matched data rather than returning a copy of the remaining input slice. This means that `buffer.match_int()` would return a `MatchedInt` _and_ consume `buffer`. When the unit test went to actually read the int from the matched bytes:
```rust
let actual = matched.read(buffer).unwrap();
```
...`buffer` would be empty, so it would fail.

I decided to match on a copy of `buffer` so the original would remain unmodified. However, `buffer` was a value, not a reference, so I couldn't simply dereference it to create a copy. I had three options:
* Make buffer a `&mut TextBuffer` that I could then dereference, adding some weirdness to the declaration in the process.
* Call `winnow`'s [`Stream::checkpoint()`](https://docs.rs/winnow/latest/winnow/stream/trait.Stream.html#tymethod.checkpoint) method, which saves the parser state (in this case, copying the `TextBuffer`), but which people won't intuitively understand.
* Call `clone()` (bleh) and live with it 'cause it's only used in the unit tests. 

I went with the latter. However, in writing this up I realized there was a fourth option that's more idiomatic to `winnow`: using [`peek()`](https://docs.rs/winnow/latest/winnow/combinator/fn.peek.html). I've updated the unit tests to use that instead.

File: src/lazy/text/raw/struct.rs [hunk 8]
```diff
@@ -1,79 +1,53 @@
...
-            }
-            Ok((_, None)) => None,
+
+        let result = whitespace_and_then(alt((
+            "}".value(None),
+            terminated(
+                E::field_expr_matcher().map(Some),
+                whitespace_and_then(opt(",")),
+            ),
+        )))
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/text/raw/struct.rs:46]: You were right to be surprised--it is indeed wrong. Similarly, _I_ was surprised to learn that neither `ion-tests` nor `ion-rust`'s test suite includes a test for a struct missing a comma:
image
🤦 

I opened https://github.com/amazon-ion/ion-tests/issues/150 to track adding a test to `ion-tests` and have added a unit test for this to `ion-rust`.

> ```
> whitespace_and_then(alt((
>     (opt(","), whitespace_and_then ("}")).value(None),
>     (",", whitespace_and_then(E::field_expr_matcher()).map(Some))
> )))
> ```

This is also not quite right since the first field cannot be preceded by a comma.

File: src/lazy/streaming_raw_reader.rs [hunk 9]
```diff
@@ -255,6 +277,9 @@ pub trait IonDataSource {
     /// Marks `number_of_bytes` in the buffer as having been read. The caller is responsible for
     /// confirming that the buffer contains at least `number_of_bytes` bytes.
     fn consume(&mut self, number_of_bytes: usize);
+
+    /// If `true`, the current contents of the buffer may not be the complete stream.
+    fn is_streaming(&self) -> bool;
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/streaming_raw_reader.rs:282]: Until you asked, I was certain that associated `const`s weren't stable on traits yet. But it works! 🎉

File: src/lazy/text/raw/sequence.rs (hunk 3 above)
Comment (zslayton) [reply src/lazy/text/raw/sequence.rs:115]: Yeah, I think so. Making it `peek` guarantees that you can keep calling `next()` on the iterator and it will still return `None`. Not critical, but a nice guard rail.

**Review by popematt — APPROVED** [review]
